### ISSUE TITLE
Backpropagate through tf.data

### ISSUE THREAD

**Issue by ufimtsev** [tensorflow/tensorflow#30672]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): n/a
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 1.14
- Python version: 3.6
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: n/a
- GPU model and memory: n/a

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
Gradients are not propagated through tf.data

**Describe the expected behavior**

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

This code snippet returns [None]:

import tensorflow as tf
_ = tf.layers.conv1d(tf.ones((1,1,2)), 1, 1, use_bias=False, name='name')
data = tf.data.Dataset.from_tensors(tf.ones((10,2))).repeat().batch(4)\
    .map(lambda x: tf.layers.conv1d(x, 1, 1, use_bias=False, name='name', reuse=True))
i = data.make_initializable_iterator()
x = i.get_next()
g = tf.gradients(x, tf.trainable_variables())
print (g)

>>> [None]

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by Uiuran** [comment]
Silly question, but did you done with Session().as_default as sess:  ->  sess.run(g)  ?

**Comment by ufimtsev** [comment]
Yes... it returns TypeError: Fetch argument None has invalid type <class 'NoneType'>

**Comment by Uiuran** [comment]
Try to retrieve the operation tensor with tensor_name = g.get_operations()[-1].name:':0', then run it...

**Comment by jsimsa** [comment]
This is not planned to be supported by the tf.data team in the near future. I will mark this issue as "contributions welcome".

**Comment by SSaishruthi** [comment]
@jsimsa 
I would like to work on this if nobody is working on this.
Is there any pointers you have that will be a good initial point for me before starting?

**Comment by jsimsa** [comment]
@SSaishruthi my suggestion would be to study the implementation of tf.gradients and extend it to tf.data. I would be happy to answer any tf.data related questions, but I am not an expert on tf.gradients.

**Comment by SSaishruthi** [comment]
Thanks @jsimsa 

I will start working on this

**Comment by noahtren** [comment]
I've been trying to implement feature map augmentation in TF2 and ran into this issue. If I augment feature maps in the middle of the network using `tf.data.`, the gradients are lost.

It seems like the best workaround is to apply operations to a batch of feature maps using Python's multiprocessing module. Has anyone tried something like this?

**Comment by Saduf2019** [comment]
@ufimtsev  
Is this still an issue.

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
